internal/repositories/rewards.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/volatiletech/sqlboiler/v4/queries/qm"
+	"github.com/volatiletech/sqlboiler/v4/types"
+)
+
+type RewardsCursor struct {
+	Week      int
+	VehicleID int
+}
+
+func RewardToAPI(reward models.Reward) gmodel.Earning {
+	stEarn := dbtypes.NullDecimalToInt(reward.StreakEarnings)
+	adEarn := dbtypes.NullDecimalToInt(reward.AftermarketEarnings)
+	syEarn := dbtypes.NullDecimalToInt(reward.SyntheticEarnings)
+
+	return gmodel.Earning{
+		Week:                    reward.IssuanceWeek,
+		Beneficiary:             common.BytesToAddress(reward.ReceivedByAddress.Bytes),
+		ConnectionStreak:        reward.ConnectionStreak.Ptr(),
+		StreakTokens:            stEarn,
+		AftermarketDeviceID:     reward.AftermarketTokenID.Ptr(),
+		AftermarketDeviceTokens: adEarn,
+		SyntheticDeviceID:       reward.SyntheticTokenID.Ptr(),
+		SyntheticDeviceTokens:   syEarn,
+		SentAt:                  reward.EarnedAt,
+		VehicleID:               reward.VehicleID,
+	}
+}
+
+func (r *Repository) GetEarningsByVehicleID(ctx context.Context, tokenID int) (*gmodel.VehicleEarnings, error) {
+	type rewardStats struct {
+		TokenSum   types.NullDecimal `boil:"token_sum"`
+		TotalCount int               `boil:"total_count"`
+	}
+	var stats rewardStats
+
+	limit := new(int)
+	*limit = 100
+
+	err := models.Rewards(
+		qm.Select(
+			fmt.Sprintf(
+				`sum(%s + %s + %s) as token_sum`,
+				models.RewardColumns.StreakEarnings, models.RewardColumns.AftermarketEarnings, models.RewardColumns.SyntheticEarnings,
+			),
+			"count(*) as total_count",
+		),
+		models.RewardWhere.VehicleID.EQ(tokenID),
+	).Bind(ctx, r.pdb.DBS().Reader, &stats)
+	if err != nil {
+		return nil, err
+	}
+
+	if stats.TokenSum.IsZero() {
+		return &gmodel.VehicleEarnings{
+			TotalTokens: &big.Int{},
+			History:     &gmodel.EarningsConnection{},
+			VehicleID:   tokenID,
+		}, nil
+	}
+
+	earningsConn := &gmodel.EarningsConnection{
+		TotalCount: stats.TotalCount,
+	}
+
+	return &gmodel.VehicleEarnings{
+		TotalTokens: dbtypes.NullDecimalToInt(stats.TokenSum),
+		History:     earningsConn,
+		VehicleID:   tokenID,
+	}, nil
+}
+
+func (r *Repository) PaginateVehicleEarningsByID(ctx context.Context, vehicleEarnings *gmodel.VehicleEarnings, first *int, after *string, last *int, before *string) (*gmodel.EarningsConnection, error) {
+	limit, err := helpers.ValidateFirstLast(first, last, maxPageSize)
+	if err != nil {
+		return nil, err
+	}
+
+	queryMods := []qm.QueryMod{
+		models.RewardWhere.VehicleID.EQ(vehicleEarnings.VehicleID),
+	}
+
+	if after != nil {
+		afterID, err := helpers.CursorToID(*after)
+		if err != nil {
+			return nil, err
+		}
+
+		queryMods = append(queryMods, models.RewardWhere.IssuanceWeek.LT(afterID))
+	}
+
+	if before != nil {
+		beforeID, err := helpers.CursorToID(*before)
+		if err != nil {
+			return nil, err
+		}
+
+		queryMods = append(queryMods, models.RewardWhere.IssuanceWeek.GT(beforeID))
+	}
+
+	orderBy := "DESC"
+	if last != nil {
+		orderBy = "ASC"
```

Comment:
We probably need to reverse later, yeah?

Lord help us when we introduce sort options.

Reply:
reverse ?

Reply:
But clients are expecting to receive these by week, descending. Specifying `first` or `last` is not supposed to change the order, just the records that are returned.

Reply:
Look at vehicles and aftermarket devices for a guide.